sympy/utilities/lambdify.py

```diff
@@ -903,6 +919,123 @@ def _module_present(modname, modlist):
             return True
     return False
 
+def _cse_homogeneous(exprs, **kwargs):
+    """
+    Same as ``cse`` but the reduced_exprs are expres in the same type as 'exprs'
+
+    Parameters
+    ==========
+
+    exprs : list of sympy expressions, or a single sympy expression
+        The expressions to reduce.
+    kwargs : The over argument of the ``cse`` function
+
+    Returns
+    =======
+
+    replacements : list of (Symbol, expression) pairs
+        All of the common subexpressions that were replaced. Subexpressions
+        earlier in this list might show up in subexpressions later in this
+        list.
+    reduced_exprs : list of sympy expressions
+        The reduced expressions with all of the replacements above.
+
+    Examples
+    ========
+
+    >>> from sympy.utilities.lambdify import _cse_homogeneous
+    >>> from sympy import Symbol, cos, Tuple
+    >>> x = Symbol("x")
+    >>> _cse_homogeneous("cos(x + 1)") # str
+    ([], 'cos(x + 1)')
+    >>> _cse_homogeneous(cos(x + 1)) # Basic
+    ([], cos(x + 1))
+    >>> _cse_homogeneous(1) # number
+    ([], 1)
+    >>> _cse_homogeneous(Tuple(1, x))
+    ([], (1, x))
+    >>> _cse_homogeneous([1, x])
+    ([], [1, x])
+    >>> _cse_homogeneous((1, x))
+    ([], (1, x))
+    >>> _cse_homogeneous({1, x})
+    ([], {1, x})
+    """
+    from sympy import cse, sympify
+
+    if isinstance(exprs, str):
+        replacements, reduced_exprs = _cse_homogeneous(sympify(exprs), **kwargs)
+        return replacements, repr(reduced_exprs)
+    if isinstance(exprs, (list, tuple, set)):
+        replacements, reduced_exprs = cse(exprs, **kwargs)
+        return replacements, type(exprs)(reduced_exprs)
+    if isinstance(exprs, dict):
+        keys = list(exprs.keys()) # In order to guarantee the order of the elements.
+        replacements, values = cse([exprs[k] for k in keys], **kwargs)
+        reduced_exprs = {k: v for k, v in zip(keys, values)}
+        return replacements, reduced_exprs
+
+    try:
+        replacements, (reduced_exprs,) = cse(exprs, **kwargs)
+    except TypeError: # For example 'mpf' objects
+        return [], exprs
+    else:
+        return replacements, reduced_exprs
+
+def _cse_del(expr):
```

Comment:
I would

- Move this helper to cse_main.py
- Make the input `replacements, reduced_expr` (i.e., the output of cse()). That way the caller can call cse() with whatever arguments they want, then process the result with this function. 
- Unless `replacements` or `reduced_exprs` are changed (they don't seem to be), this should just return `useless_symbols`.
- "useless" is a bit misleading. I would rather call this "deletable symbols".

Reply:
> * Move this helper to cse_main.py

Yes, as I said [here](https://github.com/sympy/sympy/pull/21546#issuecomment-856190288)

> * Make the input `replacements, reduced_expr` (i.e., the output of cse()). That way the caller can call cse() with whatever arguments they want, then process the result with this function.

Yes (as also shown there)

> * Unless `replacements` or `reduced_exprs` are changed (they don't seem to be), this should just return `useless_symbols`.

I disagree -- having a list of symbols that can be deleted at some point in the step-wise computation is not useful. One of the proposals here returns a set of what can be deleted at each step of the way; the version I suggested just sorts in the variables to be deleted along with the replacements and expressions. In order to do something with the symbols you need to know when they can be (manually) garbage collected.

> * "useless" is a bit misleading. I would rather call this "deletable symbols".

agree

> We should also decide if optimize is the best name for the flag. I would have called it just cse

I agree that it should not be `optimize`; `cse` is fairly neutral and as long as it works, I don't think anyone will care if you help it do garbage collection in the process by deleting variables once they are done being used.

* `cse` can handle more than one expression; should code-printing be able to return more than one expression, i.e. should lambdify be able to handle a vector of expression? If so, the code I am advocating for handles that case and allows re-ordering of expressions during calculation to try optimize when they are calculated so memory can be released. I am not guaranteeing that there is not a better way to do it, just that it can handle re-ordering.